File: cpp/src/lazy_add.cpp
#include "tide/lazy_add.h"
#include "tide/counters.h"
#include <cstdint>
#include <stdexcept>

namespace tide {
namespace {
using Ticks = std::uint64_t;
Ticks gap(Index last, Index time) {
  if (last < -1 || time < last) throw std::invalid_argument("invalid Add tick clock");
  // Avoid signed overflow even for last=-1, time=INT64_MAX.
  return (static_cast<Ticks>(time) + 1) - static_cast<Ticks>(last + 1);
}
Tensor repeat(Tensor value, const Tensor& retention, Ticks ticks) {
  for (Ticks i = 0; i < ticks; ++i) value = value * retention;
  return value;
}
class AddRepeat final : public StateKernel {
 public:
  State initial(const NodeWeights& w) const override { return {at::zeros_like(w.bias)}; }
  State step(const NodeWeights& w, const State& old, const ContentView& h, Index time) const override {
    if (time < 0 || time <= old.last_time) throw std::invalid_argument("Add requires increasing tick times");
    auto observations = increment(old.observations);
    return {h.value + repeat(old.value, w.extra.at("add_retention"), gap(old.last_time, time)), time, observations};
  }
  std::vector<State> batch(const NodeWeights& w, const std::vector<State>& old, const Tensor& h,
                           const std::vector<Index>& times, const ContentViews&) const override {
    if (old.empty() || h.dim() != 2 || h.size(0) != static_cast<Index>(old.size()) || times.size() != old.size())
      throw std::invalid_argument("invalid Add batch metadata");
    // Bucket actual samples by elapsed ticks. No padded observations or masked
    // excess decay, and no value-dependent shortcut for zero hidden/retention.
    std::map<Ticks, std::vector<Index>> groups;
    for (size_t i = 0; i < old.size(); ++i) {
      if (times[i] < 0 || times[i] <= old[i].last_time) throw std::invalid_argument("Add requires increasing tick times");
      increment(old[i].observations);
      groups[gap(old[i].last_time, times[i])].push_back(i);
    }
    std::vector<State> result(old.size());
    for (const auto& [ticks, ids] : groups) {
      std::vector<Tensor> previous, content;
      for (auto i : ids) { previous.push_back(old[i].value); content.push_back(h[i]); }
      auto values = at::stack(content) + repeat(at::stack(previous), w.extra.at("add_retention"), ticks);
      for (size_t row = 0; row < ids.size(); ++row) {
        auto i = ids[row];
        result[i] = {values[row].clone(), times[i], increment(old[i].observations)};
      }
    }
    return result;
  }
  bool joint_batch() const override { return true; }
  bool exact_sequence() const override { return true; }
  // The base sequence/packed_sequence loops preserve multiplication order and
  // report scalar_steps. This is not an associative time scan.
  void validate_weights(const NodeWeights& w) const override {
    auto it = w.extra.find("add_retention");
    if (it == w.extra.end() || !it->second.defined() || it->second.dim() != 0 || !it->second.device().is_cpu()
        || (it->second.scalar_type() != at::kFloat && it->second.scalar_type() != at::kDouble)
        || it->second.scalar_type() != w.bias.scalar_type() || it->second.device() != w.bias.device()
        || !at::isfinite(it->second).item<bool>())
      throw std::invalid_argument("Add requires finite payload-dtype scalar retention");
  }
  void validate_state(const NodeWeights&, const State& state) const override {
    if (!state.slots.empty()) throw std::invalid_argument("Add state has unexpected slots");
  }
};
}  // namespace
std::shared_ptr<const StateKernel> make_add_repeat_kernel() { return std::make_shared<AddRepeat>(); }
Tensor decode_add_repeat(const NodeWeights& w, const State& state, Index cut) {
  AddRepeat kernel; kernel.validate_weights(w); kernel.validate_state(w, state);
  if (cut < 0 || state.last_time < -1 || state.last_time >= cut || state.observations < 0)
    throw std::invalid_argument("invalid Add cut/state clock");
  if (!state.value.defined() || state.value.dim() != 1 || state.value.sizes() != w.bias.sizes()
      || state.value.scalar_type() != w.bias.scalar_type() || state.value.device() != w.bias.device()
      || !at::isfinite(state.value).all().item<bool>())
    throw std::invalid_argument("incompatible Add state value");
  return repeat(state.value, w.extra.at("add_retention"), gap(state.last_time, cut - 1));
}
}  // namespace tide

File: python/tidegraph/lazy_add.py
"""LH-compatible tick decay with explicit repeated-multiplication order."""
import torch
from .content import as_content
from .history import increment, int64
from .records import State
from .state_program import StateProgram


def repeat(value, retention, ticks):
    # No pow, affine scan or zero-value shortcut: rounding and zero VJPs matter.
    for _ in range(ticks):
        value = value * retention
    return value


class LazyAdd(StateProgram):
    profile = "lh-add-repeat-v1"
    sequence_contract = True
    joint_sequence = False  # Ordered time loop; Full/Read can still batch.

    def step(self, weights, old, content, time):
        if not int64(time) or not int64(old.last_time) or not -1 <= old.last_time < time:
            raise ValueError("Add requires strictly increasing nonnegative tick times")
        observations = increment(old.observations)
        value = repeat(old.value, weights.extra["add_retention"], time - old.last_time)
        return State(as_content(content).value + value, time, observations)

    def validate_weights(self, weights):
        rho = weights.extra.get("add_retention")
        if (rho is None or rho.ndim != 0 or rho.device.type != "cpu"
                or rho.dtype not in (torch.float32, torch.float64)
                or (rho.dtype, rho.device) != (weights.bias.dtype, weights.bias.device)
                or not torch.isfinite(rho)):
            raise ValueError("Add requires finite payload-dtype scalar retention")

    def validate(self, weights, state):
        if state.slots:
            raise ValueError("Add state has unexpected slots")


def decode(weights, state, cut):
    """Physical hidden after ticks [0,cut); no mutation or autonomous event.

    Only this explicit read incurs work for idle ticks. The stored last_time and
    observation count remain unchanged. Parameters must stay fixed for an eager
    LH interpretation across composed windows.
    """
    LazyAdd().validate_weights(weights)
    LazyAdd().validate(weights, state)
    if (not int64(cut) or not int64(state.last_time) or not int64(state.observations)
            or not -1 <= state.last_time < cut or state.observations < 0):
        raise ValueError("invalid Add cut/state clock")
    if (state.value.shape != weights.bias.shape or state.value.ndim != 1
            or (state.value.dtype, state.value.device) != (weights.bias.dtype, weights.bias.device)
            or not torch.isfinite(state.value).all()):
        raise ValueError("incompatible Add state value")
    return repeat(state.value, weights.extra["add_retention"], cut - 1 - state.last_time)
